### ISSUE TITLE
NA: make lookup return immediately

### ISSUE THREAD

**Issue by soumagne** [mercury-hpc/mercury#273]
In most cases, it is not convenient to have to make progress in order to resolve/lookup an NA address (and HG/HG core address). To get around that problem, we should think about introducing and fixing the NA layer to return immediately on lookup. Possibly the new lookup call would be:
```C
na_return_t
NA_Addr_lookup2(
        na_class_t   *na_class,
        const char   *name,
        na_addr_t   *addr
        );
```

**Comment by carns** [comment]
We don't have a general method for HG functions to return immediately do we?  Is the issue in this case that even if the lookup can be done in the context of the actual lookup call you still have to call HG_Trigger() to issue the callback for completion notification?

Just discussion to make sure there isn't some generalized solution we might want to consider.

**Comment by soumagne** [comment]
No we don't but other HG functions need progress / trigger. The goal here would be to do lookup without making progress / trigger so that users have an address that they can use immediately. That also means that potentially we'd need to defer connection until messages are sent (for plugins that need to connect) but it seems more convenient for a lookup call. Also in that case, the lookup call would no longer depend on the NA context passed and only on the NA class.

**Comment by carns** [comment]
Yeah, that's true.  I was imagining "what if a send completed immediately", but there is no raw send at the HG API level.  A forward() can't complete immediately without network progress so it doesn't need immediate completion.

So the idea is that all transports would be required to support the new immediate lookup function, and just defer connection if there is a blocking connection step?  I think that's reasonable.

**Comment by soumagne** [comment]
Right, although at the NA level it would be possible but in the send case because it can be either way, immediately completing or completing later, we would still need the callback / progress / trigger. That would not be the case for lookup anymore.

Yes all transports would need to be updated.

**Comment by carns** [comment]
Well, this isn't necessarily a good idea, but you could imagine making the semantics such that any nonblocking function is permitted to complete immediately, using some generic mechanism that would work on any HG or na_ function.  For example, just letting the function execute the callback itself directly if it manages to complete the operation during the function call.  Or have it return a special error code or something.

It might not be possible to change the semantics like that at this point, but if you could there are two possible advantages: the same approach could be applied to all functions instead of special cases, and the decision of whether to immediately complete or not could be made dynamically instead of dictated by which function variant the caller selected.

FWIW the BMI API has something a little like this.  There weren't callbacks, but any function could return 1 instead of 0 to indicate that it had completed the operation and the status was available in an output argument.  Some transports tended to complete sends immediately, others did not.  That API wasn't particularly clever though, because callers were forever having to check for 0, 1, or -errno in the return code  to figure out what to do :)

**Comment by soumagne** [comment]
They all are allowed to complete immediately currently, though the callback will only be triggered once we call NA_Trigger(). The current lookup also completes immediately in most cases, but the address can only be retrieved through the callback which is not really convenient + it is tied to the NA context being used because of the completion queue. So that change would mostly aim at fixing these two points.

**Comment by soumagne** [comment]
Merged as part of #361
